compiler/src/iree/compiler/Dialect/Util/Transforms/StripAndSplatConstants.cpp

```diff
@@ -0,0 +1,65 @@
+// Copyright 2020 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <utility>
+
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Util/Transforms/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Util {
+
+class StripAndSplatConstantsPass
+    : public StripAndSplatConstantsBase<StripAndSplatConstantsPass> {
+public:
+  StripAndSplatConstantsPass() = default;
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Util::UtilDialect>();
+  }
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
+
+    // Use a heuristic to space out splat values so it's clearer where each in
+    // the final file comes from.
+    int replaceIndex = 1;
+    auto getSplatAttr = [&](ShapedType type) {
+      return IREE::Util::BytePatternAttr::get(moduleOp.getContext(), type,
+                                              replaceIndex++);
+    };
```

Comment:
We had a heuristic in here before:

```
    // Use a heuristic to space out splat values in hopes of avoiding NaN and
    // INF values at runtime:
    //   floats: 1/1, 1/2, 1/3, ...
    //   ints: 1, 2, 3, 4, ...
```
```c++
      if (elementType.isa<FloatType>()) {
        newValue = DenseElementsAttr::get(
            tensorType, FloatAttr::get(elementType, 1.0 / replaceIndex));
      } else {
        newValue = DenseElementsAttr::get(
            tensorType, IntegerAttr::get(elementType, replaceIndex));
      }
```

This is all for easier development/debugging, so fine to just iterate based on what is observed in workloads.

IIRC, the thinking at the time was that
* if all ML model weights were positive integer values, they would eventually overflow after multiplications
* if a single value like 1 was chosen, code like `x / (x - x)` would divide by zero

Reply:
yeah, now with byte_pattern nothing can fold with these values (so no issues in the compiler) and we don't really care about execution results when all the constants are stripped (as performance and correctness are both invalid to inspect with stripped constants)